[SPARK-8124] [SPARKR]  [WIP] Created more examples on SparkR DataFrames

Emaasit opened the pull request:
Here are more examples on SparkR DataFrames including creating a Spark Contect and a SQL
context, loading data and simple data manipulation.

Emaasit committed:
Created more examples on SparkR DataFrames

Here are more examples on SparkR DataFrames including creating a SQL
context, loading data and simple data manipulation

AmplabJenkins commented:
Can one of the admins verify this patch?

Emaasit commented:
@shivaram Here is the new submission. I would like to submit a few more examples on statistical analysis and machine learning on SparkR DataFrames.

shivaram commented on examples/src/main/r/0-getting-started.R:
```diff
@@ -0,0 +1,23 @@
+#
```
We need to have the Apache License at the top of every file. You can see https://github.com/apache/spark/blob/master/examples/src/main/r/dataframe.R#L1 for an example

Also per our style guide we don't put in Author names / dates in the file itself as this is tracked in the commit log

shivaram commented on examples/src/main/r/0-getting-started.R:
```diff
@@ -0,0 +1,23 @@
+
+
+# Install SparkR from CRAN
+install.packages("SparkR")
```
In the case of these example files we will be assuming the user has installed Spark through some method (say from http://spark.apache.org/downloads.html). So we can just include the package as `library(SparkR)` and avoid the installation steps.

Emaasit committed:
Added the Apache License at the file

Emaasit committed:
Added the Apache License at the top of the file

Emaasit committed:
Added the Apache License at the top of the file

Emaasit commented:
@shivaram I have added the Apache license at the top of every file, removed author name & data.

shivaram commented on examples/src/main/r/0-getting-started.R:
```diff
@@ -0,0 +1,25 @@
+# limitations under the License.
+#
+
+# Load SparkR onto your PC
```
This comment should probably be 'Load SparkR library into your R session'

shivaram commented on examples/src/main/r/0-getting-started.R:
```diff
@@ -0,0 +1,25 @@
+sc <- sparkR.init(master = "local", appName = "MyApp")
+
+## Initialize SQLContext
+sqlCtx <- SparkRSQL.init(sc)
```
Lets use `sqlContext` as the variable name as we are using that in the sparkR shell as well

shivaram commented on examples/src/main/r/1-data.R:
```diff
@@ -0,0 +1,41 @@
+# Create an R data frame and then convert it to a SparkR DataFrame -------
+
+## Create R dataframe
+install.packages("data.table") #We want to use the fread() function to read the dataset
```
Could we use the spark-csv reader for this ? That way users will learn the way to read in large CSV files. I've done something similar in this gist here https://gist.github.com/shivaram/d0cd4aa5c4381edd6f85#file-dataframe_example-r-L6

Emaasit committed:
Updates to a comment and variable name

Now using sqlContext as the variable name

shivaram commented on examples/src/main/r/2-data-manipulation.R:
```diff
@@ -0,0 +1,62 @@
+count(flightsDF)
+
+# Show summary statistics for numeric colums
+Describe(flightsDF)
```
This should be `describe` and not `Describe` ?

shivaram commented on examples/src/main/r/2-data-manipulation.R:
```diff
@@ -0,0 +1,62 @@
+# limitations under the License.
+#
+
+source("1-data.R")
```
I'm not sure the `source` will work correctly given the way we encourage users to run example programs. If you see the `Running examples` section at http://people.apache.org/~pwendell/spark-releases/latest/index.html you can see that we ask users to run example programs with something like `./bin/spark-submit examples/src/main/r/data-manipulation.R`. 

So I'd just recommend merging all the code into one file `data-manipulation.R` and then take in the CSV file name as a command line argument.

Emaasit committed:
provided two options for creating DataFrames

provided two options for creating DataFrames. Option 1: from local data frames and option 2: directly create DataFrames using read.df function

Emaasit committed:
changed variable name to SQLContext

Emaasit committed:
combined all the code into one .R file

Deleted the source() function and combined all the code into one file

Emaasit committed:
Deleted this file

Emaasit committed:
Deleted the getting-started file

Deleted the getting started file and combined all the code into one file

Emaasit committed:
changed "Describe" function to "describe"

Emaasit committed:
Renamed file

Renamed file to data-manipulation.R

Emaasit commented:
@shivaram I wanted to provide two options for creating DataFrames. One where R users can convert their local dataframes into DataFrames and the second using the read.df().

shivaram commented on examples/src/main/r/data-manipulation.R:
```diff
@@ -0,0 +1,97 @@
+# Option 1: Create an R data frame and then convert it to a SparkR DataFrame -------
+
+## Create R dataframe
+install.packages("data.table") #We want to use the fread() function to read the dataset
```
Would `read.csv` that is part of base R also work for this ? I know that data.table is more efficient, but I would like to avoid installing new of packages in the example.

Emaasit committed:
Used base R functions

Replaced the data.table function (fread) with base R function for reading csv files (read.csv)

Emaasit commented:
@shivaram Yes, the base R function works. I have changed it.

shivaram commented on examples/src/main/r/data-manipulation.R:
```diff
@@ -0,0 +1,92 @@
+# Option 1: Create a local R data frame and then convert it to a SparkR DataFrame -------
+
+## Create a local R dataframe
+flights_df <- read.csv("flights.csv")
```
Could we take this in as a command line argument ? I think something like
```
args <- commandArgs(trailing = TRUE)
if (length(args) != 1) {
  print("Usage: data-manipulation.R <path-to-flights.csv")
  print("The data can be downloaded from: https://s3-us-west-2.amazonaws.com/sparkr-data/flights.csv ")
  q("no")
}
flightsCsvPath <- args[[1]]
```
should do the trick
```

Emaasit committed:
Using command line arguments

Taking in data set as a command line argument

Emaasit commented:
@shivaram I fixed that. You will notice that read.csv() does not work well with SSL, that is https connections. so I changed the connection to http.

shivaram commented on examples/src/main/r/data-manipulation.R:
```diff
@@ -0,0 +1,101 @@
+sc <- sparkR.init(appName = "SparkR-data-manipulation-example")
+
+## Initialize SQLContext
+sqlContext <- SparkRSQL.init(sc)
```
This should be `sparkRSQL` and not `SparkRSQL`

shivaram commented on examples/src/main/r/data-manipulation.R:
```diff
@@ -0,0 +1,101 @@
+# The data set is made up of 227,496 rows x 14 columns. 
+
+
+args <- commandArgs(trailing = TRUE)
```
Can we move this before the contexts are created right after `library(SparkR)` ? That way if users don't give an argument we will fail fast

shivaram commented on examples/src/main/r/data-manipulation.R:
```diff
@@ -0,0 +1,101 @@
+flights_df$date <- as.Date(flights_df$date)
+
+## Convert the local data frame into a SparkR DataFrame
+flightsDF <- createDataFrame(sqlContext, flights_df)
```
So I tried to run this locally and this step is very slow for the dataset we are using here (I filed https://issues.apache.org/jira/browse/SPARK-8277) due to the way we convert local data frames to lists.

I see two options here: (1) Use fewer rows in the example file, so that this runs fast or (2) use a different dataset to demonstrate creating a SparkR DataFrame from a local dataframe (the CSV reader is fine) 

Let me know which you think is better.

shivaram commented on examples/src/main/r/data-manipulation.R:
```diff
@@ -0,0 +1,101 @@
+
+# Filter flights whose destination is JFK
+jfkDF <- filter(flightsDF, "dest == JFK") ##OR
+jfkDF <- filter(flightsDF, flightsDF$dest == JFK)
```
The "JFK" should be in quotes here

shivaram commented on examples/src/main/r/data-manipulation.R:
```diff
@@ -0,0 +1,101 @@
+destDF <- sql(sqlContext, "SELECT dest, cancelled FROM flightsTable")
+
+# Use collect to create a local R data frame
+dest_df <- collect(destDF)
```
Lets name this `local_df` just to be clear ?

shivaram commented on examples/src/main/r/data-manipulation.R:
```diff
@@ -0,0 +1,101 @@
+jfkDF <- filter(flightsDF, flightsDF$dest == JFK)
+
+# Install the magrittr library
+library(magrittr)
```
Lets put this block inside a check to see if the package is installed. Something like 
```
if ("magrittr" %in% rownames(installed.packages())) {
  library(magrittr)
  ...
}
```

Emaasit committed:
Used fewer rows for createDataFrame

To create a SparkR DataFrame, I used fewer rows of the local data frame.

Emaasit commented:
@shivaram To create a Spark DataFrame from a local data frame, I used a subset of the data with fewer rows.

shivaram commented on examples/src/main/r/data-manipulation.R:
```diff
@@ -0,0 +1,101 @@
+
+# Group the flights by date and then find the average daily delay
+# Write the result into a DataFrame
+groupBy(flightsDF, "date") %>%
```
This line should also go inside the `if` block

shivaram commented:
Jenkins, ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34619 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34619/consoleFull) for   PR 6668 at commit [`3a97867`](https://github.com/apache/spark/commit/3a97867dfddd8ecd96a676b757d818f05cae4dc8).

shivaram commented on examples/src/main/r/data-manipulation.R:
```diff
@@ -0,0 +1,101 @@
+SFO_DF <- createDataFrame(sqlContext, SFO_df)
+
+#  Directly create a SparkR DataFrame from the source data
+flightsDF <- read.df(sqlContext, flightsCsvPath, source = "csv", header = "true")
```
The source here needs to be `com.databricks.spark.csv`

BTW @rxin is there some way we can map `source = csv` to that automatically ?

rxin replied:
not if csv is outside this ... maybe we can provide a way for data sources to register short names.

shivaram commented on examples/src/main/r/data-manipulation.R:
```diff
@@ -0,0 +1,101 @@
+print(local_df)
+
+# Filter flights whose destination is JFK
+jfkDF <- filter(flightsDF, "dest == JFK") ##OR
```
The `"dest == JFK"` doesn't work as the string JFK needs to be quoted. So this should be something like `jfkDF <- filter(flightsDF, "dest = \"JFK\"")`. Note that this needs to be a single `=` for things to work with this syntax.

cc @davies we should probably make `=` and `==` work here ?

davies replied:
I think we shouldn't, the expression should be either SQL or R, if we support `==` here, it's not SQL or R, will introduce other corner cases.

shivaram replied:
Ah I see - so right now the convention is that the string syntax is SQL and the normal syntax is R ?

davies replied:
Yes:)

rxin replied:
Technically we can be error tolerant and support == in the parser though. I can't think of a corner case that'd be problematic.

shivaram commented on examples/src/main/r/data-manipulation.R:
```diff
@@ -0,0 +1,101 @@
+# Group the flights by date and then find the average daily delay
+# Write the result into a DataFrame
+groupBy(flightsDF, "date") %>%
+  avg(dep_delay = "avg", arr_delay = "avg") -> dailyDelayDF
```
This needs to be `agg(dep_delay...` or to be more clear `summarize(dep_delay = ...`

shivaram commented:
Thanks @Emaasit for the update. I just had a few more things that I ran into while executing the example. Also you can verify some of these things by just running the example on your machine -- I just used a command of the form
```
./bin/sparkR --packages com.databricks:spark-csv_2.10:1.0.3 examples/src/main/r/data-manipulation.R ./flights.csv
```
to check things

SparkQA commented:
[Test build #34619 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34619/console) for   PR 6668 at commit [`3a97867`](https://github.com/apache/spark/commit/3a97867dfddd8ecd96a676b757d818f05cae4dc8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

Emaasit commented:
@shivaram Ok. Got you.

shivaram commented:
LGTM. Thanks @Emaasit for this PR. There are some outstanding comments, but I'll fix them during the merge.

Emaasit commented:
Thanks @shivaram.
